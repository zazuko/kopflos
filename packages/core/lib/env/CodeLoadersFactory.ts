import type { AnyPointer } from 'clownface'
import LoaderRegistry from 'rdf-loaders-registry'
import cjsLoader from 'rdf-loader-code/ecmaScript.js'
import esmLoader from 'rdf-loader-code/ecmaScriptModule.js'
import { isGraphPointer } from 'is-graph-pointer'

export class CodeLoadersFactory {
  private registry!: LoaderRegistry

  init() {
    this.registry = new LoaderRegistry()
    esmLoader.register(this.registry)
    cjsLoader.register(this.registry)
  }

  load<T>(code: AnyPointer): Promise<T> | T | undefined {
    if (!isGraphPointer(code)) {
      return undefined
    }

    return this.registry.load<T>(code)
  }

  static exports = ['load']
}
