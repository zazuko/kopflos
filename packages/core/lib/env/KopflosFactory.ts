import type { KopflosConfig } from '../Kopflos.js'

export interface KopflosFactory {
  readonly kopflos: {
    readonly config: KopflosConfig
  }
}

export default (config: KopflosConfig) => class implements KopflosFactory {
  get kopflos() {
    return {
      config: Object.freeze(config),
    }
  }

  static exports = ['kopflos']
}
