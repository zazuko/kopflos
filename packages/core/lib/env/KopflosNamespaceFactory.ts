import type { NamespaceBuilder } from '@rdfjs/namespace'
import type { NamespaceFactory } from '@rdfjs/namespace/Factory.js'
import type { Environment } from '@rdfjs/environment/Environment.js'
import type NsBuildersFactory from '@tpluscode/rdf-ns-builders'
import { code } from '@zazuko/vocabulary-extras-builders'

type KopflosTerms = 'api' | 'Config' | 'Api' | 'resourceLoader' | 'handler' | 'method'

declare module '@tpluscode/rdf-ns-builders' {
  interface CustomNamespaces {
    kopflos: NamespaceBuilder<KopflosTerms>
    code: typeof code
  }
}

export class KopflosNamespaceFactory {
  init(this: Environment<NamespaceFactory | NsBuildersFactory>) {
    this.ns = {
      ...this.ns,
      kopflos: this.namespace('https://kopflos.described.at/'),
      code,
    }
  }
}
