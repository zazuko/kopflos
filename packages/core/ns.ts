import rdf from '@zazuko/env-node'

type Properties = 'api' | 'resourceLoader' | 'handler' | 'method' | 'config'
type Classes = 'Config' | 'Api' | 'ResourceShape' | 'Handler'
type Shorthands = 'DescribeLoader' | 'OwnGraphLoader'

export interface KopflosTerms extends Record<Properties | Classes | Shorthands, never>{
}

export const kl = rdf.namespace<keyof KopflosTerms>('https://kopflos.described.at/')
