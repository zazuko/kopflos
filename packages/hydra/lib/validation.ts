import type { HandlerArgs, KopflosEnvironment } from '@kopflos-cms/core'
import { isGraphPointer, isNamedNode } from 'is-graph-pointer'
import type { ShapesGraphLoader } from '@kopflos-cms/shacl'
import type { MultiPointer } from 'clownface'
import type { DatasetCore } from '@rdfjs/types'

export const shapesGraphLoader: ShapesGraphLoader = ({ env, subject }: HandlerArgs) => {
  const { sh, owl, kl, hydra, rdf } = env.ns

  let memberShape = subject.out(kl('hydra#memberCreateShape'))
  if (!isGraphPointer(memberShape)) {
    memberShape = subject.out(kl('hydra#memberShape'))
  }

  const shapesGraph = env.clownface()
  copySubgraph(env, memberShape, shapesGraph.dataset)

  shapesGraph
    .blankNode()
    .addOut(rdf.type, sh.NodeShape)
    .addOut(sh.targetNode, subject)
    .addOut(sh.property, prop => {
      prop
        .addOut(sh.path, hydra.member)
        .addOut(sh.maxCount, 1)
        .addOut(sh.minCount, 1)
        .addOut(sh.message, 'Posting to a collection requires a single new hydra:member')
    })

  shapesGraph
    .blankNode()
    .addOut(sh.targetObjectsOf, hydra.member)
    .addOut(sh.node, memberShape)

  const imports = subject
    .out(sh.shapesGraph)
    .filter(isNamedNode)
  shapesGraph
    .blankNode()
    .addOut(owl.imports, imports)

  return shapesGraph.dataset
}

function copySubgraph(env: KopflosEnvironment, source: MultiPointer, destination: DatasetCore) {
  const traverser = env.traverser<DatasetCore>(({ level, quad: { subject } }) => {
    return level === 0 || subject.termType === 'BlankNode'
  })

  source.toArray()
    .forEach(ptr => {
      traverser.forEach(ptr, ({ quad }) => {
        destination.add(quad)
      })
    })
}
