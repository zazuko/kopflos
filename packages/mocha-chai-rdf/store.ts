/* eslint-disable no-console,camelcase */
import * as url from 'node:url'
import assert from 'node:assert'
import * as Oxigraph from 'oxigraph'
import type { NamespaceBuilder } from '@rdfjs/namespace'
import rdf from '@zazuko/env-node'
import type { DatasetCore, NamedNode, Quad_Graph } from '@rdfjs/types'
import type { AnyPointer } from 'clownface'
import type { Dataset } from '@zazuko/env/lib/Dataset.js'

declare module 'mocha' {
  interface Context {
    dataset: DatasetCore
    graph: AnyPointer
    store: Oxigraph.Store
  }
}

interface DatasetSourceOptions {
  format?: 'trig' | 'nq'
  loadAll?: boolean
  includeDefaultGraph?: boolean
}

interface GraphSourceOptions {
  format?: 'ttl' | 'nt'
}

type Options = (DatasetSourceOptions | GraphSourceOptions) & {
  baseIri?: string | NamespaceBuilder
}

export function createStore(base: string, options: Options = {}) {
  const format = options.format ?? 'ttl'
  let loadAll = true
  let includeDefaultGraph = false
  if (options.format === 'trig' || options.format === 'nq') {
    loadAll = options.loadAll || false
    includeDefaultGraph = options.includeDefaultGraph || false
  }

  return async function (this: Mocha.Context) {
    const store = new Oxigraph.Store()

    const path = url.fileURLToPath(new url.URL(`${base}.${format}`))

    let dataset: Dataset = await rdf.dataset().import(rdf.fromFile(path))

    let graph: Quad_Graph | undefined
    if (this.currentTest && !loadAll) {
      graph = testGraph(this.currentTest)
      dataset = dataset
        .filter((quad) => {
          return quad.graph.equals(graph) || (includeDefaultGraph && quad.graph.equals(rdf.defaultGraph()))
        })
        .map(function toDefaultGraph(quad) {
          return rdf.quad(quad.subject, quad.predicate, quad.object)
        })
    }

    for (const quad of dataset) {
      store.add(quad)
    }

    function assertNotEmpty() {
      if (loadAll) {
        return assert(dataset.size > 0, 'Test data not found')
      }
      if (graph && !graph.equals(rdf.defaultGraph())) {
        return assert(
          dataset.size > 0,
          `Test data not found in GRAPH <${graph.value}>`,
        )
      }

      return assert(
        dataset.match(null, null, null, graph).size > 0,
        `Test data not found in GRAPH <${graph?.value}>`,
      )
    }

    Object.defineProperty(this, 'dataset', {
      get() {
        assertNotEmpty()
        return dataset
      },
      configurable: true,
    })

    Object.defineProperty(this, 'graph', {
      get() {
        assertNotEmpty()
        return rdf.clownface({ dataset })
      },
      configurable: true,
    })
    Object.defineProperty(this, 'store', {
      get() {
        assertNotEmpty()
        return store
      },
      configurable: true,
    })

    // hack to remove the graph after the test just at the right moment
    const cleanup = this.test?.title.includes('before all') ? after : afterEach
    cleanup(() => {
      /* eslint-disable @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      delete this.dataset
      // @ts-ignore
      delete this.graph
      // @ts-ignore
      delete this.store
    })
  }
}

function testGraph(test: Mocha.Test): NamedNode {
  return rdf.namedNode(encodeURI(test.titlePath().slice(1, -1).map(removeSpaces).join('/')))
}

function removeSpaces(arg: string) {
  return arg.replaceAll(' ', '-')
}
