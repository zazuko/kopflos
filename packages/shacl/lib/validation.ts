import type { ValidationReport } from 'rdf-validate-shacl/src/validation-report.js'
import SHACLEngine from 'rdf-validate-shacl'
import type { HandlerArgs } from '@kopflos-cms/core'
import type { Options } from '../index.js'
import { findShapes, loadShapes } from './shapes.js'

export interface ValidationCallback {
  (args: HandlerArgs, loadDataGraph: Options['loadDataGraph']): Promise<ValidationReport>
}

export async function shaclValidate(args: HandlerArgs, loadDataGraph: Options['loadDataGraph'] = dataGraphFromPayload): Promise<ValidationReport> {
  // TODO: do not call findShapes twice
  const shapes = await findShapes(args)
  const shapesGraph = await loadShapes(shapes, args.env)
  const engine = new SHACLEngine(shapesGraph, { factory: args.env })

  return engine.validate(await loadDataGraph({
    ...args,
    shapesGraph,
  }))
}

function dataGraphFromPayload({ body }: HandlerArgs) {
  return body.dataset
}
