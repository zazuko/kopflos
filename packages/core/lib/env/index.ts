import parent, { Environment } from '@zazuko/env-node'
import { KopflosNamespaceFactory } from './KopflosNamespaceFactory.js'
import { CodeLoadersFactory } from './CodeLoadersFactory.js'

const env = new Environment([
  KopflosNamespaceFactory,
  CodeLoadersFactory,
], { parent })

export default env
export type KopflosEnvironment = typeof env
