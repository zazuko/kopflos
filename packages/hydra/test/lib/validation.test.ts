import { expect } from 'chai'
import type { HandlerArgs, KopflosEnvironment } from '@kopflos-cms/core'
// eslint-disable-next-line import/no-unresolved
import { createEnv } from '@kopflos-cms/core/env.js'
import { createStore } from 'mocha-chai-rdf/store.js'
import { shapesGraphLoader } from '../../lib/validation.js'
import inMemoryClients from '../../../testing-helpers/in-memory-clients.js'
import { ex } from '../../../testing-helpers/ns.js'

describe('@kopflos-cms/hydra/lib/validation.js', function () {
  let env: KopflosEnvironment

  beforeEach(createStore(import.meta.url, {
    format: 'trig',
  }))

  beforeEach(function () {
    env = createEnv({
      baseIri: 'http://example.org/',
      sparql: {
        default: inMemoryClients(this.rdf),
      },
    })
  })

  describe('shapesGraphLoader', function () {
    context('collection has hydra#memberCreateShape', function () {
      it('copies it to output shapes', async function () {
        // given
        const args = <HandlerArgs>{
          env,
          subject: this.rdf.graph.node(ex.collection),
        }

        // when
        const shapesGraph = await shapesGraphLoader(args)

        // then
        expect(shapesGraph).canonical.toMatchSnapshot()
      })
    })
    context('collection has hydra#createMemberCreateShape', function () {
      it('copies it to output shapes', async function () {
        // given
        const args = <HandlerArgs>{
          env,
          subject: this.rdf.graph.node(ex.collection),
        }

        // when
        const shapesGraph = await shapesGraphLoader(args)

        // then
        expect(shapesGraph).canonical.toMatchSnapshot()
      })
    })

    context('when collection has sh:shapesGraph', function () {
      it('includes referenced sh:shapesGraph', async function () {
        // given
        const args = <HandlerArgs>{
          env,
          subject: this.rdf.graph.node(ex.collection),
        }

        // when
        const shapesGraph = await shapesGraphLoader(args)

        // then
        const imports = [...shapesGraph.match(null, env.ns.owl.imports)]
        expect(imports.map((m) => m.object)).to.deep.contain.all.members([
          ex.fooShapes,
          ex.barShapes,
        ])
      })
    })
  })
})
