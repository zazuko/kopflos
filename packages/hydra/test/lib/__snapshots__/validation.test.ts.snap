// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@kopflos-cms/hydra/lib/validation.js shapesGraphLoader collection has hydra#createMemberCreateShape copies it to output shapes 1`] = `
"_:c14n0 <http://www.w3.org/ns/shacl#node> _:c14n3 .
_:c14n0 <http://www.w3.org/ns/shacl#targetObjectsOf> <http://www.w3.org/ns/hydra/core#member> .
_:c14n1 <http://www.w3.org/ns/shacl#hasValue> <http://schema.org/Person> .
_:c14n1 <http://www.w3.org/ns/shacl#path> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/shacl#NodeShape> .
_:c14n2 <http://www.w3.org/ns/shacl#property> _:c14n4 .
_:c14n2 <http://www.w3.org/ns/shacl#targetNode> <http://example.org/collection> .
_:c14n3 <http://www.w3.org/ns/shacl#property> _:c14n1 .
_:c14n4 <http://www.w3.org/ns/shacl#maxCount> \\"1\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
_:c14n4 <http://www.w3.org/ns/shacl#message> \\"Posting to a collection requires a single new hydra:member\\" .
_:c14n4 <http://www.w3.org/ns/shacl#minCount> \\"1\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
_:c14n4 <http://www.w3.org/ns/shacl#path> <http://www.w3.org/ns/hydra/core#member> .
"
`;

exports[`@kopflos-cms/hydra/lib/validation.js shapesGraphLoader collection has hydra#memberCreateShape copies it to output shapes 1`] = `
"_:c14n0 <http://www.w3.org/ns/shacl#node> _:c14n3 .
_:c14n0 <http://www.w3.org/ns/shacl#targetObjectsOf> <http://www.w3.org/ns/hydra/core#member> .
_:c14n1 <http://www.w3.org/ns/shacl#hasValue> <http://schema.org/Person> .
_:c14n1 <http://www.w3.org/ns/shacl#path> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> .
_:c14n2 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://www.w3.org/ns/shacl#NodeShape> .
_:c14n2 <http://www.w3.org/ns/shacl#property> _:c14n4 .
_:c14n2 <http://www.w3.org/ns/shacl#targetNode> <http://example.org/collection> .
_:c14n3 <http://www.w3.org/ns/shacl#property> _:c14n1 .
_:c14n4 <http://www.w3.org/ns/shacl#maxCount> \\"1\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
_:c14n4 <http://www.w3.org/ns/shacl#message> \\"Posting to a collection requires a single new hydra:member\\" .
_:c14n4 <http://www.w3.org/ns/shacl#minCount> \\"1\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
_:c14n4 <http://www.w3.org/ns/shacl#path> <http://www.w3.org/ns/hydra/core#member> .
"
`;
